zerver/actions/message_send.py

```diff
@@ -293,14 +311,31 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
         )
         followed_topic_push_user_ids = followed_topic_notification_recipients("push_notifications")
 
+        if possible_stream_wildcard_mention or possible_topic_wildcard_mention:
+            # We calculate `wildcard_mentions_notify_user_ids` and `followed_topic_wildcard_mentions_notify_user_ids`
+            # only if there's a possible stream or topic wildcard mention in the message.
+            # This is important so as to avoid unnecessarily sending huge user ID lists with
+            # thousands of elements to the event queue (which can happen because these settings
+            # are `True` by default for new users.)
+            wildcard_mentions_notify_user_ids = notification_recipients("wildcard_mentions_notify")
+            followed_topic_wildcard_mentions_notify_user_ids = (
+                followed_topic_notification_recipients("wildcard_mentions_notify")
+            )
+
         if possible_stream_wildcard_mention:
-            # We calculate `stream_wildcard_mention_user_ids` and `followed_topic_wildcard_mention_user_ids`
-            # only if there's a possible stream wildcard mention in the message. This is important so as
-            # to avoid unnecessarily sending huge user ID lists with thousands of elements to the
-            # event queue (which can happen because these settings are `True` by default for new users.)
-            stream_wildcard_mention_user_ids = notification_recipients("wildcard_mentions_notify")
+            stream_wildcard_mention_user_ids = wildcard_mentions_notify_user_ids
             stream_wildcard_mention_in_followed_topic_user_ids = (
-                followed_topic_notification_recipients("wildcard_mentions_notify")
+                followed_topic_wildcard_mentions_notify_user_ids
```

Comment:
This code looks much cleaner than the first version!

One question I have is, should we be doing a deep/shallow copy of the common variables (`wildcard_mentions_notify_user_ids` and `followed_topic_wildcard_mentions_notify_user_ids`)? I don't remember how Python does copying of variables etc, but I do remember than its a bit different than other languages like C++/Java.

If not that, we should at least make sure that we have a test simulating a scenario of a message containing both stream and topic mentions, and then verify that our user IDs sets are the same as what we expect.

Reply:
Oh I see we have a test for that, great! :)

Reply:
Had a discussion around this with Abhijeet on CZO (DM):

Here it makes a shallow copy. But even if the sets are changed at a later stage, it will not affect the behavior as we are no longer going to reuse the `followed_topic_wildcard_mentions_notify_user_ids` or `wildcard_mentions_notify_user_ids`.

And we already have tests to cover the expected behavior. So, we're good to go.